### PR TITLE
Fix/c tape

### LINKED ISSUE #574: make Construction Markers render like fences
Currently construction markers are two blocks and have a saved orientation.

We could improve this by rendering them like fences. One block, one orientation and rendering is dynamically done by checking the surrounding. This would make them feel more minecrafty and allow players to use them somewhere else.

<bountysource-plugin>

---
Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/41892952-make-construction-markers-render-like-fences?utm_campaign=plugin&utm_content=tracker%2F43175463&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F43175463&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by Raycoms**
Closes #574

# Changes proposed in this pull request:
- Improves Ctape code to only need one block class

Review please

### CODE REVIEW

**Review by marchermans — APPROVED** [review]

File: src/main/java/com/minecolonies/coremod/blocks/BlockConstructionTape.java [hunk 1]
```diff
@@ -276,6 +439,86 @@ public IBlockState getStateForPlacement(
...
     protected BlockStateContainer createBlockState()
     {
-        return new BlockStateContainer(this, FACING);
+        return new BlockStateContainer(this, new IProperty[] {FACING, VARIANT});
+    }
+
+    /**
+     * Types that the {@link BlockConstructionTape} supports
+     */
+    public enum Type implements IStringSerializable
```
Comment (BuildTeamLinter) [inline src/main/java/com/minecolonies/coremod/blocks/BlockConstructionTape.java:448]: MAJOR Reduce this class from 59 to the maximum allowed 30 or externalize it in a public class. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS2972)

File: src/main/java/com/minecolonies/coremod/blocks/BlockConstructionTape.java [hunk 2]
```diff
@@ -182,8 +270,48 @@ public boolean isPassable(final IBlockAccess worldIn, final BlockPos pos)
     @SuppressWarnings(DEPRECATION)
     @Override
     @Deprecated
-    public AxisAlignedBB getBoundingBox(final IBlockState state, final IBlockAccess source, final BlockPos pos)
+    public AxisAlignedBB getBoundingBox(final IBlockState stateIn, final IBlockAccess source, final BlockPos pos)
```
Comment (BuildTeamLinter) [inline src/main/java/com/minecolonies/coremod/blocks/BlockConstructionTape.java:273]: INFO Do not forget to remove this deprecated code someday. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1133)

File: src/main/java/com/minecolonies/coremod/blocks/BlockConstructionTape.java [hunk 3]
```diff
@@ -205,6 +333,37 @@ public AxisAlignedBB getBoundingBox(final IBlockState state, final IBlockAccess
         }
     }
 
+    /**
+     * @deprecated remove when minecraft invents something better.
+     */
+    @Deprecated
+    @Override
+    public IBlockState getActualState(@NotNull final IBlockState state, @NotNull final IBlockAccess worldIn, @NotNull final BlockPos pos)
```
Comment (BuildTeamLinter) [inline src/main/java/com/minecolonies/coremod/blocks/BlockConstructionTape.java:341]: INFO Do not forget to remove this deprecated code someday. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1133)

File: src/main/java/com/minecolonies/coremod/blocks/BlockConstructionTape.java [hunk 4]
```diff
@@ -205,6 +333,37 @@ public AxisAlignedBB getBoundingBox(final IBlockState state, final IBlockAccess
...
+     * @return the blockState to use.
+     */
+    private static IBlockState getTapeShape(@NotNull final IBlockState state, @NotNull final IBlockAccess world, @NotNull final BlockPos position)
+    {
+        final boolean[] connectors = new boolean[]{world.getBlockState(position.east()).getBlock() instanceof BlockConstructionTape,
+                world.getBlockState(position.west()).getBlock() instanceof BlockConstructionTape,
+                world.getBlockState(position.north()).getBlock() instanceof BlockConstructionTape,
+                world.getBlockState(position.south()).getBlock() instanceof BlockConstructionTape};
+
+        if((connectors[0] && connectors[2]) || (connectors[0] && connectors[3]) || (connectors[1] && connectors[3]) || (connectors[1] && connectors[2]))
```
Comment (BuildTeamLinter) [inline src/main/java/com/minecolonies/coremod/blocks/BlockConstructionTape.java:360]: CRITICAL Reduce the number of conditional operators (7) used in the expression (maximum allowed 6). [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1067)
MAJOR Assign this magic number 3 to a well-named constant, and use the constant instead. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS109)
MAJOR Assign this magic number 3 to a well-named constant, and use the constant instead. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS109)

File: src/main/java/com/minecolonies/coremod/blocks/BlockConstructionTape.java [hunk 5]
```diff
@@ -255,9 +414,13 @@ public BlockRenderLayer getBlockLayer()
...
     }
 
+    /**
+     * @deprecated (Remove this as soon as minecraft offers anything better).
+     */
     @SuppressWarnings(DEPRECATION)
     @NotNull
     @Override
+    @Deprecated
     public IBlockState getStateForPlacement(
```
Comment (BuildTeamLinter) [inline src/main/java/com/minecolonies/coremod/blocks/BlockConstructionTape.java:424]: INFO Do not forget to remove this deprecated code someday. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1133)

**Comment by BuildTeamLinter** [comment]
SonarQube analysis reported 56 issues
* ![CRITICAL][CRITICAL] 1 critical
* ![MAJOR][MAJOR] 45 major
* ![MINOR][MINOR] 1 minor
* ![INFO][INFO] 9 info

Watch the comments in this conversation to review them.

#### Top 10 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MAJOR][MAJOR] [blocks#](https://github.com/Minecolonies/minecolonies/blob/aac4534a5369268c64c3688f787080e09282a311/src/main/java/com/minecolonies/coremod/blocks#): Add a 'package-info.java' file to document the 'blocks' package [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)
1. ![MAJOR][MAJOR] [ModBlocks.java#L15](https://github.com/Minecolonies/minecolonies/blob/aac4534a5369268c64c3688f787080e09282a311/src/main/java/com/minecolonies/coremod/blocks/ModBlocks.java#L15): Split this class into smaller and more specialized ones to reduce its dependencies on other classes from 35 to the maximum authorized 30 or less. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1200)
1. ![MAJOR][MAJOR] [gui#](https://github.com/Minecolonies/minecolonies/blob/aac4534a5369268c64c3688f787080e09282a311/src/main/java/com/minecolonies/coremod/client/gui#): Add a 'package-info.java' file to document the 'gui' package [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)
1. ![MAJOR][MAJOR] [render#](https://github.com/Minecolonies/minecolonies/blob/aac4534a5369268c64c3688f787080e09282a311/src/main/java/com/minecolonies/coremod/client/render#): Add a 'package-info.java' file to document the 'render' package [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)
1. ![MAJOR][MAJOR] [colony#](https://github.com/Minecolonies/minecolonies/blob/aac4534a5369268c64c3688f787080e09282a311/src/main/java/com/minecolonies/coremod/colony#): Add a 'package-info.java' file to document the 'colony' package [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)
1. ![MAJOR][MAJOR] [buildings#](https://github.com/Minecolonies/minecolonies/blob/aac4534a5369268c64c3688f787080e09282a311/src/main/java/com/minecolonies/coremod/colony/buildings#): Add a 'package-info.java' file to document the 'buildings' package [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)
1. ![MAJOR][MAJOR] [permissions#](https://github.com/Minecolonies/minecolonies/blob/aac4534a5369268c64c3688f787080e09282a311/src/main/java/com/minecolonies/coremod/colony/permissions#): Add a 'package-info.java' file to document the 'permissions' package [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)
1. ![MAJOR][MAJOR] [requestable#](https://github.com/Minecolonies/minecolonies/blob/aac4534a5369268c64c3688f787080e09282a311/src/main/java/com/minecolonies/coremod/colony/requestable#): Add a 'package-info.java' file to document the 'requestable' package [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)
1. ![MAJOR][MAJOR] [init#](https://github.com/Minecolonies/minecolonies/blob/aac4534a5369268c64c3688f787080e09282a311/src/main/java/com/minecolonies/coremod/colony/requestsystem/init#): Add a 'package-info.java' file to document the 'init' package [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)
1. ![MAJOR][MAJOR] [locations#](https://github.com/Minecolonies/minecolonies/blob/aac4534a5369268c64c3688f787080e09282a311/src/main/java/com/minecolonies/coremod/colony/requestsystem/locations#): Add a 'package-info.java' file to document the 'locations' package [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)

[CRITICAL]: https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL'
[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'
[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Pull request merged**